File: logcat_viewer/filters.py
"""Filter specification and matching, logcat-style (level / tag / PID / regex text)."""
from __future__ import annotations

import re
from dataclasses import dataclass, field

from .parser import LogEntry, PRIORITY


@dataclass
class FilterSpec:
    """User-facing filter settings. Call compile() once, then match() per entry."""

    min_priority: int = PRIORITY["V"]      # keep entries with priority >= this
    tag_query: str = ""
    tag_regex: bool = False
    pids: str = ""                         # comma/space separated PIDs; empty = any
    package: str = ""                      # selected app (label only)
    package_pids: "frozenset[int] | None" = None  # None = no app filter; else pid must be in set
    text_query: str = ""                   # include: tag+msg must match
    text_regex: bool = False
    exclude_query: str = ""                # exclude: hide if tag+msg matches
    exclude_regex: bool = False

    # populated by compile()
    _tag_re: "re.Pattern | None" = field(default=None, init=False, repr=False)
    _text_re: "re.Pattern | None" = field(default=None, init=False, repr=False)
    _exclude_re: "re.Pattern | None" = field(default=None, init=False, repr=False)
    # substring OR-terms (lowercased, split on "|"); empty list => field inactive
    _tag_terms: "list[str]" = field(default_factory=list, init=False, repr=False)
    _text_terms: "list[str]" = field(default_factory=list, init=False, repr=False)
    _exclude_terms: "list[str]" = field(default_factory=list, init=False, repr=False)
    _pid_set: "frozenset[int]" = field(default=frozenset(), init=False, repr=False)
    errors: "dict[str, str]" = field(default_factory=dict, init=False, repr=False)

    def compile(self) -> "FilterSpec":
        """Precompile regexes / parse PIDs. Invalid regex in a field disables that
        field (pass-through) and records an error message under its key."""
        self.errors = {}
        self._tag_re, self._tag_terms = self._compile_field(
            "tag", self.tag_query, self.tag_regex)
        self._text_re, self._text_terms = self._compile_field(
            "text", self.text_query, self.text_regex)
        self._exclude_re, self._exclude_terms = self._compile_field(
            "exclude", self.exclude_query, self.exclude_regex)

        pids = set()
        for tok in re.split(r"[\s,]+", self.pids.strip()):
            if tok.isdigit():
                pids.add(int(tok))
            elif tok:
                self.errors["pid"] = f"not a PID: {tok}"
        self._pid_set = frozenset(pids)
        return self

    def _compile_field(self, key: str, query: str, is_regex: bool):
        """Return (compiled_regex_or_None, substring_terms). In substring mode a
        query is split on "|" into OR-terms — `error|success` matches lines with
        either word; empty terms (e.g. a trailing `|`) are ignored."""
        if not query:
            return None, []
        if not is_regex:
            terms = [t for t in (p.strip().lower() for p in query.split("|")) if t]
            return None, terms
        try:
            return re.compile(query, re.IGNORECASE), []
        except re.error as exc:
            self.errors[key] = str(exc)
            return None, []  # invalid -> treat field as inactive so log keeps flowing

    def has_error(self, key: str) -> bool:
        return key in self.errors

    def match(self, e: LogEntry) -> bool:
        if e.priority < self.min_priority:
            return False

        if self.package_pids is not None and e.pid not in self.package_pids:
            return False

        if self._pid_set and e.pid not in self._pid_set:
            return False

        if self._tag_re is not None:
            if not self._tag_re.search(e.tag):
                return False
        elif self._tag_terms:  # substring OR: keep if the tag has ANY term
            tag = e.tag.lower()
            if not any(t in tag for t in self._tag_terms):
                return False

        if self._text_re is not None:
            if not self._text_re.search(e.raw):
                return False
        elif self._text_terms:  # substring OR: keep if a term is in tag+message
            if not any(t in e.search for t in self._text_terms):
                return False

        if self._exclude_re is not None:
            if self._exclude_re.search(e.raw):
                return False
        elif self._exclude_terms:  # substring OR: hide if ANY term is present
            if any(t in e.search for t in self._exclude_terms):
                return False

        return True

File: androidlab-electron/src/core/filters.ts
/**
 * Filter specification and matching, logcat-style (level / tag / PID / regex text).
 * Faithful port of logcat_viewer/filters.py.
 *
 * Regex note: user-entered patterns are compiled with JavaScript's RegExp (the
 * `i` flag == Python's re.IGNORECASE). The vast majority of logcat filter
 * patterns are syntactically identical across the two engines; the handful of
 * Python-only constructs (e.g. inline `(?P<name>)`) are the accepted, documented
 * difference. Invalid patterns disable that one field and record an error,
 * exactly like the Python version — the stream keeps flowing.
 */
import type { LogEntry } from './parser'
import { PRIORITY } from './priorities'

export interface FilterSpecInit {
  minPriority?: number
  tagQuery?: string
  tagRegex?: boolean
  pids?: string
  packageName?: string
  packagePids?: ReadonlySet<number> | null
  textQuery?: string
  textRegex?: boolean
  excludeQuery?: string
  excludeRegex?: boolean
}

export class FilterSpec {
  minPriority: number
  tagQuery: string
  tagRegex: boolean
  pids: string
  packageName: string
  packagePids: ReadonlySet<number> | null
  textQuery: string
  textRegex: boolean
  excludeQuery: string
  excludeRegex: boolean

  // populated by compile()
  private _tagRe: RegExp | null = null
  private _textRe: RegExp | null = null
  private _excludeRe: RegExp | null = null
  private _tagTerms: string[] = []
  private _textTerms: string[] = []
  private _excludeTerms: string[] = []
  private _pidSet: Set<number> = new Set()
  errors: Record<string, string> = {}

  constructor(init: FilterSpecInit = {}) {
    this.minPriority = init.minPriority ?? PRIORITY.V
    this.tagQuery = init.tagQuery ?? ''
    this.tagRegex = init.tagRegex ?? false
    this.pids = init.pids ?? ''
    this.packageName = init.packageName ?? ''
    this.packagePids = init.packagePids ?? null
    this.textQuery = init.textQuery ?? ''
    this.textRegex = init.textRegex ?? false
    this.excludeQuery = init.excludeQuery ?? ''
    this.excludeRegex = init.excludeRegex ?? false
  }

  /**
   * Precompile regexes / parse PIDs. Invalid regex in a field disables that
   * field (pass-through) and records an error message under its key.
   */
  compile(): this {
    this.errors = {}
    ;[this._tagRe, this._tagTerms] = this._compileField('tag', this.tagQuery, this.tagRegex)
    ;[this._textRe, this._textTerms] = this._compileField('text', this.textQuery, this.textRegex)
    ;[this._excludeRe, this._excludeTerms] = this._compileField(
      'exclude',
      this.excludeQuery,
      this.excludeRegex
    )

    const pids = new Set<number>()
    for (const tok of this.pids.trim().split(/[\s,]+/)) {
      if (/^\d+$/.test(tok)) {
        pids.add(parseInt(tok, 10))
      } else if (tok) {
        this.errors.pid = `not a PID: ${tok}`
      }
    }
    this._pidSet = pids
    return this
  }

  /**
   * Return [compiledRegexOrNull, substringTerms]. In substring mode a query is
   * split on "|" into OR-terms — `error|success` matches lines with either
   * word; empty terms (e.g. a trailing `|`) are ignored.
   */
  private _compileField(key: string, query: string, isRegex: boolean): [RegExp | null, string[]] {
    if (!query) {
      return [null, []]
    }
    if (!isRegex) {
      const terms = query
        .split('|')
        .map((p) => p.trim().toLowerCase())
        .filter((t) => t.length > 0)
      return [null, terms]
    }
    try {
      return [new RegExp(query, 'i'), []]
    } catch (exc) {
      this.errors[key] = exc instanceof Error ? exc.message : String(exc)
      return [null, []] // invalid -> treat field as inactive so log keeps flowing
    }
  }

  hasError(key: string): boolean {
    return key in this.errors
  }

  match(e: LogEntry): boolean {
    if (e.priority < this.minPriority) {
      return false
    }

    if (this.packagePids !== null && !this.packagePids.has(e.pid)) {
      return false
    }

    if (this._pidSet.size > 0 && !this._pidSet.has(e.pid)) {
      return false
    }

    if (this._tagRe !== null) {
      if (!this._tagRe.test(e.tag)) {
        return false
      }
    } else if (this._tagTerms.length > 0) {
      // substring OR: keep if the tag has ANY term
      const tag = e.tag.toLowerCase()
      if (!this._tagTerms.some((t) => tag.includes(t))) {
        return false
      }
    }

    if (this._textRe !== null) {
      if (!this._textRe.test(e.raw)) {
        return false
      }
    } else if (this._textTerms.length > 0) {
      // substring OR: keep if a term is in tag+message
      if (!this._textTerms.some((t) => e.search.includes(t))) {
        return false
      }
    }

    if (this._excludeRe !== null) {
      if (this._excludeRe.test(e.raw)) {
        return false
      }
    } else if (this._excludeTerms.length > 0) {
      // substring OR: hide if ANY term is present
      if (this._excludeTerms.some((t) => e.search.includes(t))) {
        return false
      }
    }

    return true
  }
}
